File: weisi/src/main/java/com/cjq/lib/weisi/communicator/receiver/DataReceiver.java
package com.cjq.lib.weisi.communicator.receiver;

import com.cjq.lib.weisi.communicator.Communicator;

import java.io.IOException;
import java.util.concurrent.Executor;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * Created by CJQ on 2017/12/22.
 */

public class DataReceiver {

    private static final int DEFAULT_BUFFER_SIZE = 2048;

    private final Communicator mCommunicator;
    //private volatile boolean mListening;
    private State mState;
    private Listener mListener;

    public DataReceiver(Communicator communicator) {
        if (communicator == null) {
            throw new NullPointerException("communicator may not be null");
        }
        mCommunicator = communicator;
        mState = State.STOPPED;
    }

    public boolean startListen(Listener listener) {
        return startListen(listener, DEFAULT_BUFFER_SIZE);
    }

    public synchronized boolean startListen(Listener listener, final int bufferSize) {
        if (mState == State.STOPPED) {
            if (listener == null || bufferSize <= 0 || !mCommunicator.canRead()) {
                return false;
            }
            mListener = listener;
            mState = State.RUNNING;
            new Thread(new Runnable() {
                @Override
                public void run() {
                    int receivedLen = 0;
                    int handledLen;
                    byte[] data = new byte[bufferSize];
                    while (isListening() && mCommunicator.canRead()) {
                        try {
                            receivedLen += mCommunicator.read(data, receivedLen, data.length - receivedLen);
                            if (receivedLen > 0) {
                                handledLen = mListener.onDataReceived(data, receivedLen);
                                receivedLen = saveUnhandledData(data, receivedLen, handledLen);
                            }
                        } catch (IOException ioe) {
                            if (mListener.onErrorOccurred(ioe)) {
                                mState = State.STOPPING;
                            }
                        }
                    }
                    mState = State.STOPPED;
                    mListener = null;
                }
            }).start();
        }
        return isListening();
    }

    private int saveUnhandledData(byte[] data, int receivedLen, int handledLen) {
        if (handledLen >= receivedLen) {
            return 0;
        } else if (handledLen > 0 && handledLen < receivedLen) {
            for (int i = handledLen; i < receivedLen; ++i) {
                data[i - handledLen] = data[i];
            }
            return receivedLen - handledLen;
        } else {
            return receivedLen;
        }
    }

    public boolean isListening() {
        return mState == State.RUNNING;
    }

    public synchronized void stopListen() {
        if (isListening()) {
            mState = State.STOPPING;
            onStopListen(mCommunicator);
        }
    }

    protected void onStopListen(Communicator communicator) {
    }

    protected State getState() {
        return mState;
    }

    protected enum State {
        STOPPED,
        RUNNING,
        STOPPING
    }

    public interface Listener {
        //len为实际接收数据长度
        //返回已处理字节数
        int onDataReceived(byte[] data, int len);
        //返回true表示停止监听，返回false表示继续监听
        boolean onErrorOccurred(Exception e);
    }
}

File: wsb/src/main/java/com/wsn/lib/wsb/communicator/receiver/DataReceiver.kt
package com.wsn.lib.wsb.communicator.receiver


import com.wsn.lib.wsb.communicator.Communicator

import java.io.IOException

/**
 * Created by CJQ on 2017/12/22.
 */

open class DataReceiver(private val communicator: Communicator) {

    protected var state: State = State.STOPPED
        private set
    //private var listener: Listener? = null

    val isListening: Boolean
        get() = state === State.RUNNING

    fun startListen(listener: Listener): Boolean {
        return startListen(listener, DEFAULT_BUFFER_SIZE)
    }

    @Synchronized
    fun startListen(listener: Listener, bufferSize: Int): Boolean {
        if (state == State.STOPPED) {
            if (bufferSize <= 0 || !communicator.canRead()) {
                return false
            }
            //listener = l
            state = State.RUNNING
            Thread(Runnable {
                var receivedLen: Int
                var handledLen: Int
                var bufferLen = 0
                val data = ByteArray(bufferSize)
                while (isListening && communicator.canRead()) {
                    try {
                        receivedLen = communicator.read(data, bufferLen, data.size - bufferLen)
                        if (receivedLen > 0) {
                            bufferLen += receivedLen
                            handledLen = listener.onDataReceived(data, bufferLen)
                            bufferLen = saveUnhandledData(data, bufferLen, handledLen)
                        } else if (receivedLen == -1) {
                            throw ReceiveException("recv len = 1")
                        }
                    } catch (ioe: IOException) {
                        if (listener.onErrorOccurred(ioe)) {
                            state = State.STOPPING
                        }
                    }
                }
                state = State.STOPPED
            }).start()
        }
        return isListening
    }

    private fun saveUnhandledData(data: ByteArray, receivedLen: Int, handledLen: Int): Int {
        if (handledLen >= receivedLen) {
            return 0
        } else if (handledLen > 0 && handledLen < receivedLen) {
            for (i in handledLen until receivedLen) {
                data[i - handledLen] = data[i]
            }
            return receivedLen - handledLen
        } else {
            return receivedLen
        }
    }

    @Synchronized
    fun stopListen() {
        if (isListening) {
            state = State.STOPPING
            communicator.stopRead()
        }
    }

    //protected open fun onStopListen(communicator: Communicator) {}

    protected enum class State {
        STOPPED,
        RUNNING,
        STOPPING
    }

    interface Listener {
        //len为实际接收数据长度
        //返回已处理字节数
        fun onDataReceived(data: ByteArray, len: Int): Int

        //返回true表示停止监听，返回false表示继续监听
        fun onErrorOccurred(e: Exception): Boolean
    }

    companion object {

        private const val DEFAULT_BUFFER_SIZE = 2048
    }
}
